src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java

```diff
@@ -0,0 +1,298 @@
+import org.apache.commons.lang.Validate;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.knn.common.exception.OutOfNativeMemoryException;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.plugin.stats.StatNames;
+
+import java.io.Closeable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Manages native memory allocations made by JNI.
+ */
+public final class NativeMemoryCacheManager implements Closeable {
+
+    public static String GRAPH_COUNT = "graph_count";
+
+    private static Logger logger = LogManager.getLogger(NativeMemoryCacheManager.class);
+    private static NativeMemoryCacheManager INSTANCE;
+
+    private Cache<String, NativeMemoryAllocation> cache;
+    private ExecutorService executor;
+    private AtomicBoolean cacheCapacityReached;
+    private long maxWeight;
+
+    NativeMemoryCacheManager() {
+        this.executor = Executors.newSingleThreadExecutor();
+        this.cacheCapacityReached = new AtomicBoolean(false);
+        this.maxWeight = Long.MAX_VALUE;
+        initialize();
+    }
+
+    /**
+     * Make sure we just have one instance of cache.
+     *
+     * @return NativeMemoryCacheManager instance
+     */
+    public static synchronized NativeMemoryCacheManager getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new NativeMemoryCacheManager();
+        }
+        return INSTANCE;
+    }
+
+    private void initialize() {
+        CacheBuilder<String, NativeMemoryAllocation> cacheBuilder = CacheBuilder.newBuilder()
+                .recordStats()
+                .concurrencyLevel(1)
+                .removalListener(this::onRemoval);
+
+        if(KNNSettings.state().getSettingValue(KNNSettings.KNN_MEMORY_CIRCUIT_BREAKER_ENABLED)) {
+            maxWeight = KNNSettings.getCircuitBreakerLimit().getKb();
+            cacheBuilder.maximumWeight(maxWeight).weigher((k, v) -> (int)v.getSizeInKb());
+        }
+
+        if(KNNSettings.state().getSettingValue(KNNSettings.KNN_CACHE_ITEM_EXPIRY_ENABLED)) {
+            long expiryTime = ((TimeValue) KNNSettings.state()
+                    .getSettingValue(KNNSettings.KNN_CACHE_ITEM_EXPIRY_TIME_MINUTES)).getMinutes();
+            cacheBuilder.expireAfterAccess(expiryTime, TimeUnit.MINUTES);
+        }
+
+        cacheCapacityReached = new AtomicBoolean(false);
+
+        cache = cacheBuilder.build();
+    }
+
+    /**
+     * Evicts all entries from the cache and rebuilds.
+     */
+    public synchronized void rebuildCache() {
+        logger.info("KNN Cache rebuilding.");
+        executor.execute(() -> {
+            cache.invalidateAll();
```

Comment:
You are invalidating all elements at 98 but creating new cache in next step. is this expected?

Reply:
Right, I think we explicitly want to invalidate the entries in the old cache so that any allocations get dealloc'd.

I am not sure if this would happen anyway when the old cache is garbage collected, but this is how it is currently done in the plugin, so I rather not make a change to something that works.